### ISSUE TITLE
The `manual_find` lint provides a wrong suggestion if the try operator is used

### ISSUE THREAD

**Issue by mkrasnitski** [rust-lang/rust-clippy#9521]
### Summary

Because the try operator cannot be used in closures, replacing a manual implementation of find with an actual function call results in changed semantics that are not simple to resolve.

### Lint Name

manual_find

### Reproducer

I tried this code:

```rust
fn assert_less_ne(x: u32, y: u32) -> Result<bool, ()> {
    if x == y {
        return Err(());
    }
    Ok(x < y)
}

fn first_less_than_10(v: Vec<u32>) -> Option<u32> {
    for x in v {
        if assert_less_ne(x, 10).ok()? {
            return Some(x)
        }
    }
    None
}
```

I saw this happen:

```
warning: manual implementation of `Iterator::find`
  --> src/main.rs:9:5
   |
9  | /     for x in v {
10 | |         if assert_less_ne(x, 10).ok()? {
11 | |             return Some(x);
12 | |         }
13 | |     }
14 | |     None
   | |________^ help: replace with an iterator: `v.into_iter().find(|&x| assert_less_ne(x, 10).ok()?)`
   |
   = note: `#[warn(clippy::manual_find)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_find
```

I expected to see this happen:

Either the help text should appear with no suggestion or a modified suggestion, or the lint should not be triggered at all.

### Version

```text
rustc 1.64.0 (a55dd71d5 2022-09-19)
binary: rustc
commit-hash: a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52
commit-date: 2022-09-19
host: x86_64-unknown-linux-gnu
release: 1.64.0
LLVM version: 14.0.6
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by kraktus** [comment]
For some reasons `clippy_utils::can_move_expr_to_closure` is overly conservative and using It would create more false negative, such as

```diff
 fn with_struct(arr: Vec<Data>) -> Option<Data> {
-    arr.into_iter().find(|el| el.name.len() == 10)
+    for el in arr {
+        if el.name.len() == 10 {
+            return Some(el);
+        }
+    }
+    None
 }
 ```
